### PR TITLE
Add Default Tag to New Functions

### PR DESCRIPTION
**Pull request by james-r-smith**
Add `latest` tag to new functions created by OpenFaaS CLI tool.

## Description

Append `latest` tag to all newly created functions. This fixes a bug
where images being pushed without a tag will update all tags of the
image. Now, only the current tag will be pushed. Fix is implemented with
Golang template to allow for greater flexibility.

Fixes: openfaas/faas#779

Signed-off-by: James Smith <jamessmithsmusic@gmail.com>

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

References [[#779](https://github.com/openfaas/faas/issues/779)] in [openfaas/faas](https://github.com/openfaas/faas).

## How Has This Been Tested?

Existing tests were ran successfully.

Verified existing functions without a tag are not broken by :

- Creating new function
- removing supplied tag `latest`
- Build / Push function
- Update tag to `v0.0.1`
- Build / Push function
- verify that only tag `v0.0.1` is updated
- Remove tag
- verify that both tags were updated

Verified new functionality by :

- Creating new function
  - template tag used is `latest`
- Build / Push function
- Update tag to `v0.0.1`
- Build / Push function
- verify that only tag `v0.0.1` is updated
- Update tag to `latest`
- verify that only tag `latest` is updated

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
  - Modified current tests to satisfy.
  - Added a few additional tests to `new` command for further coverage
- [x] All new and existing tests passed.

### CODE REVIEW

File: commands/new_function.go [hunk 1]
```diff
@@ -123,91 +123,97 @@ func runNewFunction(cmd *cobra.Command, args []string) error {
 			return fmt.Errorf("unable to find file: %s - %s", appendFile, statErr.Error())
 		}
 
-		duplicateError := duplicateFunctionName(functionName, appendFile)
+		var duplicateError error
+		services, duplicateError = duplicateFunctionName(functionName, appendFile)
```
Comment (james-r-smith) [inline commands/new_function.go]: @alexellis would you like me to add an additional commit here to cover ensuring past functions get `latest` tag if they aren't tagged? I was going to add this, but I thought the validation could possibly be done elsewhere and hence left for a potential debate. Just let me know. 😊👍

File: commands/new_function.go [hunk 2]
```diff
@@ -123,91 +123,97 @@ func runNewFunction(cmd *cobra.Command, args []string) error {
...
+	} else {
+		fmt.Printf("Stack file written: %s\n", fileName)
 	}
 
 	if !quiet {
 		languageTemplate, _ := stack.LoadLanguageTemplate(language)
 
 		if languageTemplate.WelcomeMessage != "" {
-			fmt.Println("\nNotes:")
+			fmt.Printf("\nNotes:\n")
```
Comment (alexellis) [inline commands/new_function.go]: Don't change it in this PR, but this could be one `Printf` for both lines

File: commands/new_function.go [hunk 3]
```diff
@@ -217,37 +223,32 @@ functions:
 
 func printAvailableTemplates(availableTemplates []string) string {
 	var result string
-	sort.Sort(StrSort(availableTemplates))
+	sort.Slice(availableTemplates, func(i, j int) bool {
```
Comment (alexellis) [inline commands/new_function.go]: Just inlined the sorter? We lost the intent though.. (sort by string), might need comment adding in.

File: commands/new_function.go [hunk 4]
```diff
@@ -217,37 +223,32 @@ functions:
...
+		return availableTemplates[i] < availableTemplates[j]
+	})
 	for _, template := range availableTemplates {
 		result += fmt.Sprintf("- %s\n", template)
 	}
 	return result
 }
 
-func duplicateFunctionName(functionName string, appendFile string) error {
+func duplicateFunctionName(functionName string, appendFile string) (*stack.Services, error) {
```
Comment (alexellis) [inline commands/new_function.go]: Save for follow-up PR, but I think the read should happen outside the function `duplicateFunctionName` - we shouldn't want methods doing more than one thing.

**Comment by alexellis** [comment]
Works great, thanks for this PR @james-r-smith 

```bash
alext:tmp alex$ /Users/alex/go/src/github.com/openfaas/faas-cli/faas-cli new --lang go go-james
2018/07/31 09:53:16 No templates found in current directory.
2018/07/31 09:53:16 Attempting to expand templates from https://github.com/openfaas/templates.git
2018/07/31 09:53:17 Fetched 13 template(s) : [csharp dockerfile go go-armhf java8 node node-arm64 node-armhf python python-armhf python3 python3-armhf ruby] from https://github.com/openfaas/templates.git
Folder: go-james created.
  ___                   _____           ____
 / _ \ _ __   ___ _ __ |  ___|_ _  __ _/ ___|
| | | | '_ \ / _ \ '_ \| |_ / _` |/ _` \___ \
| |_| | |_) |  __/ | | |  _| (_| | (_| |___) |
 \___/| .__/ \___|_| |_|_|  \__,_|\__,_|____/
      |_|


Function created in folder: go-james
Stack file written: go-james.yml

Notes:
You have created a new function which uses Golang 1.9.7.
To include third-party dependencies, use a vendoring tool like dep:
dep documentation: https://github.com/golang/dep#installation

alext:tmp alex$ cat go-james.yml 
provider:
  name: faas
  gateway: http://127.0.0.1:8080

functions:
  go-james:
    lang: go
    handler: ./go-james
    image: go-james:latest
```

**Pull request merged**